### PR TITLE
Update Microsoft.CodeAnalysis.Analyzers to 3.3.4

### PR DESCRIPTION
**Pull request by Youssef1313**

### CODE REVIEW

**Comment by Youssef1313** [comment]
```
##[error]src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs(83,40): error RS1035: (NETCORE_ENGINEERING_TELEMETRY=Build) The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables
/mnt/vss/_work/1/s/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs(331,45): error RS1035: The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables [/mnt/vss/_work/1/s/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj::TargetFramework=netstandard2.0]
##[error]src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs(331,45): error RS1035: (NETCORE_ENGINEERING_TELEMETRY=Build) The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables

```

@mavasani Are these false positives?

**Comment by mavasani** [comment]
> ```
> ##[error]src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs(83,40): error RS1035: (NETCORE_ENGINEERING_TELEMETRY=Build) The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables
> /mnt/vss/_work/1/s/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs(331,45): error RS1035: The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables [/mnt/vss/_work/1/s/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj::TargetFramework=netstandard2.0]
> ##[error]src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs(331,45): error RS1035: (NETCORE_ENGINEERING_TELEMETRY=Build) The symbol 'Environment' is banned for use by analyzers: Analyzers should not read their settings directly from environment variables
> ```
> 
> @mavasani Are these false positives?

@Youssef1313 We should be setting the EnforceExtendedAnalyzerRules property to false for `Microsoft.CodeAnalysis` project, I don't believe we want RS1035 and RS1036 to be enabled for it. @RikkiGibson to confirm

File: eng/config/globalconfigs/Common.globalconfig [hunk 1]
```diff
@@ -59,6 +59,9 @@ dotnet_diagnostic.RS0046.severity = suggestion
 # RS0047: (Suppressor) The 'Type.FullName' property is never null when using the syntax 'typeof(T).FullName'
 dotnet_diagnostic.RS0047.severity = none
 
+# RS1035: Do not use APIs banned for analyzers
+dotnet_diagnostic.RS1035.severity = none
```
Comment (mavasani) [inline eng/config/globalconfigs/Common.globalconfig:63]: We should be disabling RS1036 not RS1035, which is already disabled by default.

File: eng/targets/Settings.props [hunk 2]
```diff
@@ -49,6 +49,7 @@
 
     <!-- VS does not require Windows PDBs to be published anymore. -->
     <PublishWindowsPdb>false</PublishWindowsPdb>
+    <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
```
Comment (mavasani) [inline eng/targets/Settings.props:52]: This doesn't look correct - the property should default to `false`. This property should be set to `true` only for projects with analyzers or generators, i.e. Features layer and Code Style layer projects for analyzers.

File: eng/config/globalconfigs/Common.globalconfig (hunk 1 above)
Comment (Youssef1313) [inline eng/config/globalconfigs/Common.globalconfig:63]:
```suggestion
# RS1014 - Re-enable when we have a fix from https://github.com/dotnet/roslyn-analyzers/pull/6442
dotnet_diagnostic.RS1014.severity = none
```

File: eng/targets/Settings.props (hunk 2 above)
Comment (Youssef1313) [inline eng/targets/Settings.props:52]:
```suggestion
    <EnforceExtendedAnalyzerRules>false</EnforceExtendedAnalyzerRules>
```

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 3]
```diff
@@ -2501,7 +2501,9 @@ internal bool CreateDebugDocuments(DebugDocumentsBuilder documentsBuilder, IEnum
             // Therefore we need to add these in a separate pass after documents for syntax trees were added.
             foreach (var tree in CommonSyntaxTrees)
             {
+#pragma warning disable RS1014 // https://github.com/dotnet/roslyn-analyzers/pull/6442 - remove suppression when we consume analyzers with the fix.
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2504]:
```suggestion
```

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 4]
```diff
@@ -2501,7 +2501,9 @@ internal bool CreateDebugDocuments(DebugDocumentsBuilder documentsBuilder, IEnum
...
                 AddDebugSourceDocumentsForChecksumDirectives(documentsBuilder, tree, diagnostics);
+#pragma warning restore RS1014
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2506]:
```suggestion
```

**Comment by Youssef1313** [comment]
> ##[error]src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs(1793,57): error RS1037: (NETCORE_ENGINEERING_TELEMETRY=Build) Add "CompilationEnd" custom tag to the diagnostic descriptor used to initialize field 'SymbolStartedEndedDifferRule' as it is used to report a compilation end diagnostic


@mavasani Do we care for compiler test projects?

**Comment by mavasani** [comment]
> > ##[error]src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs(1793,57): error RS1037: (NETCORE_ENGINEERING_TELEMETRY=Build) Add "CompilationEnd" custom tag to the diagnostic descriptor used to initialize field 'SymbolStartedEndedDifferRule' as it is used to report a compilation end diagnostic
> 
> @mavasani Do we care for compiler test projects?

Nope, this custom tag only matters for error list scenarios, which can be pragma suppressed for analyzers used in unit tests.

**Comment by mavasani** [comment]
> > > ##[error]src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs(1793,57): error RS1037: (NETCORE_ENGINEERING_TELEMETRY=Build) Add "CompilationEnd" custom tag to the diagnostic descriptor used to initialize field 'SymbolStartedEndedDifferRule' as it is used to report a compilation end diagnostic
> > 
> > 
> > @mavasani Do we care for compiler test projects?
> 
> Nope, this custom tag only matters for error list scenarios, which can be pragma suppressed for analyzers used in unit tests.

Probably even better to disable the rule in non-shipping globalconfig file for the repo.

File: src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs [hunk 5]
```diff
@@ -4,6 +4,8 @@
 
 #nullable disable
 
+#pragma warning disable RS1037 // Add "CompilationEnd" custom tag to compilation end diagnostic descriptor - we don't care for tests.
+
```
Comment (Youssef1313) [inline src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs:8]:
```suggestion
```

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Looks like only our "Fixes" project has IsAnalyzer=true, and not any "Analyzers" project. Is that fine?

https://cs.github.com/dotnet/roslyn?q=IsAnalyzer+language%3AXML

**Comment by Youssef1313** [comment]
@RikkiGibson This project is referencing the analyzer projects:

https://github.com/dotnet/roslyn/blob/1903beb881257b1e51632885595b0c4ff6863e4a/src/CodeStyle/CSharp/CodeFixes/Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes.csproj#L54-L56

I'm not sure if `IsAnalyzer` property is supposed to flow when building the project dependencies? I think it does?

**Comment by RikkiGibson** [comment]
I'm not sure it would. It makes sense for analyzers to depend on things which aren't considered analyzers. So I wouldn't expect it to happen automatically.

Basically, it might be worth making a pass over projects which "are definitely just analyzers/generators/fixers". But it doesn't necessarily need to happen in this PR.

**Comment by Youssef1313** [comment]
Yeah I inspected a binlog from artifacts and it indeed doesn't appear to happen automatically.

I'll open an issue.

**Comment by Youssef1313** [comment]
@RikkiGibson Is this ready to merge?

**Review by jaredpar — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Thanks @Youssef1313!

**Pull request merged**